Feat/subjects

Anurag-Wednesday opened the pull request:
### Ticket Link

---

### Related Links

---

### Description

1. Created method to handle responses when no results are found.
2. Updated the getAllSubjects handler
3. Wrote tests for the subjects route
4. Created new findOne and findAll route for the user_subjects
5. added cacheDrop for the patch and delete operations.

### Steps to Reproduce / Test

---

---

### Checklist

- [ ] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added]
- [ ] Relevant documentation is changed or added (and PR referenced)

### GIF's

---
![2022-02-16 19 16 09](https://user-images.githubusercontent.com/98798828/154277516-f77922eb-2ef2-469a-85df-88f4878fe707.gif)

alichherawalla committed:
user_subjects

Anurag-Wednesday committed:
Updated route "/subjects"

1. Created method to handle responses when no results are found.
2. Updated the getAllSubjects handler
3. Wrote tests for the subjects route

Anurag-Wednesday committed:
Updated the Routes for subjects and UserSubkjects

1. Created route to getAllSubjects
2. Create route to add subjects
3. Created route to delete and update subjects
4. Created route to get one UserSubject
5. Created route to get all userSubjects

Anurag-Wednesday committed:
reset cache on update and delete operations

Added dropCache to make sure that the cache doesn't serve stale information

Anurag-Wednesday committed:
added more routest to user_subjects

1. Created new findOne and findAll route for the user_subjects
2. added cacheDrop for the patch and delete operations.

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+        order: sequelize.literal('id ASC'),
+        limit,
+        offset: (page - 1) * limit,
+        include: [
```
let's pass "include" as a parameter to the dao

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+            id
+        }
+    });
+    const get = await findOneSubject(id);
```
let's please rename

alichherawalla replied:
this would be more appropriate. 
Trying to update something that doesn't exist plus looking it up isn't ideal
```
export const updateOneSubject = async (id, updatedSubject) => {
    const subject = await findOneSubject(id);
    if (subject) {
        await subject.update(updatedSubject);
        return { name: subject.name, id: subject.id };
    } else {
        return badRequest('Did not update anything, please check the Id');
    }
};
```

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+
+export const addOneSubject = async subject => {
+    try {
+        const added = await subjects.create(subject);
```
better variable names please

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+    }
+};
+
+export const addOneSubject = async subject => {
```
better variable names please

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+    }
+};
+
+export const deleteOneSubject = async id => {
```
If we've deleted it we shouldn't be able to fetch it from the DB again

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,61 @@
+    if (await findOneSubject(id)) {
+        return `Deleted item by id ${id}`;
+    } else {
+        return badRequest('Did not delete anything, please check the Id');
```
Id -> id

alichherawalla commented on lib/daos/tests/subjectsDao.test.js:
```diff
@@ -0,0 +1,102 @@
+        it('should find a subject by ID', async () => {
+            const { findOneSubject } = require('daos/subjectsDAO');
+            const testSubject = await findOneSubject(1);
+            expect(testSubject.id).toEqual(1);
```
expect(testSubject.id).toEqual(mockSubject.name);

alichherawalla commented on lib/daos/tests/subjectsDao.test.js:
```diff
@@ -0,0 +1,102 @@
+            const { findAllSubjects } = require('daos/subjectsDao');
+            const { allSubjects } = await findAllSubjects(1, 10);
+            const firstSubject = allSubjects[0];
+            expect(firstSubject.id).toEqual(1);
```
test against if of mockSubject

alichherawalla commented on lib/daos/tests/userSubjectsDAO.test.js:
```diff
@@ -0,0 +1,21 @@
+import { mockData } from 'utils/mockData';
+describe('userSubject daos', () => {
```
let's please keep the naming consistent for all the files. 
```
userSubjectsDao.test.js
```

alichherawalla commented on lib/daos/tests/userSubjectsDAO.test.js:
```diff
@@ -0,0 +1,21 @@
+        it('should find a userSubject by ID', async () => {
+            const { findOneUserSubject } = require('daos/userSubjectsDao');
+            const testSubject = await findOneUserSubject(1);
+            expect(testSubject.id).toEqual(1);
```
mockSubject.id. This is try for all. If we are reusing this in a lot of places probably makes sense to rollup in a function

alichherawalla commented on lib/daos/userSubjectsDao.js:
```diff
@@ -0,0 +1,21 @@
+        include
+    });
+
+    return {
```
it would make sense for the dao layer to spit out a consistent format. 

For example since we already know that the hapi-pagination plugin requires results and totalCount we should return that from the daos that are returning an array. 

results, totalCount

In this way you don't need to reassign in the route

alichherawalla commented on lib/routes/subjects/routes.js:
```diff
@@ -0,0 +1,157 @@
+                .then(subjects => {
+                    if (get(subjects.allSubjects, 'length')) {
+                        const totalCount = subjects.totalCount;
+                        const allSubjects =
```
move this to the dao

alichherawalla commented on lib/routes/subjects/routes.js:
```diff
@@ -0,0 +1,157 @@
+                                subjects.allSubjects
+                            ).map(subject => subject);
+
+                        return h.response({
```
the dao should return as ```{ results, totalCount }```

alichherawalla commented on lib/routes/subjects/routes.js:
```diff
@@ -0,0 +1,157 @@
+        handler: async request => {
+            const subjectId = request.params.subjectId;
+            const name = request.payload.name;
+            return server.methods.findOneSubject.cache
```
updateOneSubject function throws an error if the id doesn't match a valid subject, we would still be dropping the cache. This would be an unnecessary write to the cache

alichherawalla commented on lib/routes/subjects/routes.js:
```diff
@@ -0,0 +1,157 @@
+        },
+        handler: async request => {
+            const name = request.payload.name;
+            return addOneSubject({ name });
```
we typically use insertOneSubject

alichherawalla commented on lib/routes/subjects/routes.js:
```diff
@@ -0,0 +1,157 @@
+        },
+        handler: async request => {
+            const subjectId = request.params.subjectId;
+            return server.methods.findOneSubject.cache
```
same thing here for when the cache is getting cleared

alichherawalla commented on lib/routes/user-subjects/routes.js:
```diff
@@ -0,0 +1,82 @@
+                    if (get(userSubjects.allUserSubjects, 'length')) {
+                        const totalCount = userSubjects.totalCount;
+                        const allUserSubjects =
+                            transformDbArrayResponseToRawResponse(
```
please move these to the dao

Anurag-Wednesday committed:
made changes as per the comments

1. Updated the names of some variables
2. fixed issues with extra write to cache
3. Updated the Dao and routes structure

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,64 @@
+    return notFound('No subjects found');
+};
+
+export const updateOneSubject = async (id, subject) => {
```
```
export const updateOneSubject = async (id, updatedSubject) => {
    let subject = await findOneSubject(id);
    if (subject) {
       subject = await subject.update(updatedSubject);
       return {id: subject.id, name: subject.name}
    }
    return badRequest('Did not update anything, please check the id');
}
```

Anurag-Wednesday replied:
fixed the updateOneSubject and deleteOneSubject functions

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,64 @@
+    }
+};
+
+export const insertOneSubject = async subject => {
```
insertSubject

Anurag-Wednesday replied:
Updated the name

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,64 @@
+
+export const insertOneSubject = async subject => {
+    try {
+        const insertedSubject = await subjects.create(subject);
```
```
return subjects.create(subject).catch( ... )
```

alichherawalla commented on lib/daos/subjectsDao.js:
```diff
@@ -0,0 +1,64 @@
+};
+
+export const deleteOneSubject = async id => {
+    if (await findOneSubject(id)) {
```
destroy on the model

Anurag-Wednesday committed:
made changes as per the comments

1. Updated the name for InsertOneSubject
2. refactored the functions for updateOne and deleteOne subject
